boot/freeldr/bootsect/ntfs.S

```diff
@@ -0,0 +1,890 @@
+
+    mov  dx, es
+    add  dx, 32                                 // Increment read buffer for next sector
+    mov  es, dx
+
+    loop ReadSectorsCHSLoop                     // Read next sector
+
+    pop es
+    ret
+
+// Displays a disk error message
+// And reboots
+PrintDiskError:
+    mov  si, offset msgDiskError        // Bad boot disk message
+    call PutChars                       // Display it
+
+    jmp  Reboot
+
+// Displays a file system error message
+// And reboots
+PrintFileSystemError:
+    mov  si, offset msgFileSystemError  // FreeLdr not found message
+    call PutChars                       // Display it
+
+Reboot:
+    mov  si, offset msgAnyKey           // Press any key message
+    call PutChars                       // Display it
+    xor  ax, ax
+    int  HEX(16)                        // Wait for a keypress
+    int  HEX(19)                        // Reboot
+
+PutChars:
+    lodsb
+    or   al, al
+    jz   short Done
+    mov  ah, HEX(0e)
+    mov  bx, 7
+    int  HEX(10)
+    jmp  short PutChars
+Done:
+    ret
+
+msgDiskError:
+    .ascii "Disk error", CR, LF, NUL
+msgFileSystemError:
+    .ascii "File system error", CR, LF, NUL
+msgAnyKey:
+    .ascii "Press any key to restart", CR, LF, NUL
+
+SectsPerMFT:
+    .word 0
+    
+MFTStartSector:
+    .quad 0
+
+BiosCHSDriveSize:
+    .long 0
+
+.org 509 // Pad to 509 bytes
+BootPartition:
+    .byte 1
+
+BootFlagSignature:
+    .word HEX(0aa55)    // BootSector signature
+
+// End of Bootsector
+// Next sector starts as sector 2, since boot sector is 
+// as a file under NTFS ($Boot, which has inode number 7)
+// and takes the two first clusters of the volume (which means
+// 16 sectors, checked on Windows and mkfs.ntfs from ntfsutils)
+
+#define ATTRIBUTE_DATA HEX(80)
+#define ATTRIBUTE_INDEX_ROOT HEX(90)
+#define ATTRIBUTE_INDEX_ALLOCATION HEX(A0)
+#define FILE_MAGIC HEX(454c4946)
+#define INDX_MAGIC HEX(58444e49)
+StartSearch:
+    // Compute MFT start sector
+    mov   eax, dword ptr [MftLocation]
+    mov   cl, byte ptr [SectsPerCluster]
+    movzx ecx, cx
+    mul   ecx
+    mov   dword ptr [MFTStartSector], eax
+
+    // Compute size of MFT entry in sectors
+    xor   ax, ax
+    mov   al, byte ptr [ClustersPerMftRecord]
+    test  al, al
+    js    NegativeOffset
+    mov   cx, word ptr [SectsPerCluster]
+    mul   cx
+    mov   word ptr [SectsPerMFT], cx
+    jmp   SearchNext
+    NegativeOffset:
+    // ClustersPerMftRecord is negative, so we need to perform 1 << (-ClustersPerMftRecord)
+    // to get the number of bytes needed for a MFT entry
+    not   al
+    inc   al
+    mov   cl, al
+    xor   ax, ax
+    inc   ax
+    shl   ax, cl
+
+    // But here want to store sectors, so we need to divide by BytesPerSector
+    xor   dx, dx
+    mov   cx, word ptr [BytesPerSector]
+    div   cx
+    mov   word ptr [SectsPerMFT], ax
+
+    SearchNext:
+    mov   bp, sp
+    sub   sp, HEX(10)
+
+    // Read Root Directory MFT
+    mov   ax, HEX(2000)
```

Comment:
Can you document the location and size of the disk buffer? See https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/bootsect/ext2.S;h=bd3119c635feafba68ee36677098b4aa3efdea0d;hb=refs/heads/master#l129